Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionSidebarCocoa.mm

```diff
@@ -0,0 +1,299 @@
+#import "WebExtensionSidebar.h"
+
+#if ENABLE(WK_WEB_EXTENSIONS_SIDEBAR)
+
+#import "WebExtensionContext.h"
+#import "WebExtensionTab.h"
+#import "WebExtensionWindow.h"
+
+namespace WebKit {
+
+static NSString * const sidebarActionManifestKey = @"sidebar_action";
+static NSString * const sidePanelManifestKey = @"side_panel";
+static NSString * const sidebarActionTitleKey = @"default_title";
+static NSString * const sidebarActionPathKey = @"default_panel";
+static NSString * const sidePanelPathKey = @"default_path";
+
+static NSString * const fallbackPath = @"about:blank";
+static NSString * const fallbackTitle = @"";
+
+static std::optional<RetainPtr<NSDictionary<NSString *, id>>> ensureDict(id maybeDict)
+{
+    if (!maybeDict)
+        return std::nullopt;
+    if (![maybeDict isKindOfClass:NSDictionary.class])
+        return std::nullopt;
+    return (NSDictionary<NSString *, id> *) maybeDict;
+}
+
+static std::optional<String> ensureStringValue(RetainPtr<NSDictionary<NSString *, id>> dict, RetainPtr<NSString> key)
+{
+    if (!dict || !key)
+        return std::nullopt;
+    id maybeValue = [dict.get() objectForKey:key.get()];
+    if (![maybeValue isKindOfClass:NSString.class])
+        return std::nullopt;
+    return (NSString *) maybeValue;
+}
+
+static std::optional<String> getDefaultSidebarTitleFromExtension(WebExtension& extension)
+{
+    id maybeSidebarAction = extension.manifest()[sidebarActionManifestKey];
+
+    // since sidePanel does not have a default title specified in manifest, we only check sidebarAction
+    return ensureDict(maybeSidebarAction)
+        .and_then([](RetainPtr<NSDictionary<NSString *, id>> sidebarAction) { return ensureStringValue(sidebarAction.get(), sidebarActionTitleKey); })
+        .value_or(extension.displayName());
+}
+
+static std::optional<String> getDefaultSidebarPathFromExtension(WebExtension& extension)
+{
+    NSDictionary *manifest = extension.manifest();
+    auto maybePanelTitle = ensureDict(manifest[sidebarActionManifestKey])
+        .and_then([](RetainPtr<NSDictionary<NSString *, id>> sidebarAction) {
+            return ensureStringValue(sidebarAction.get(), sidebarActionPathKey);
+        });
+
+    // since std::optional::or_else cannot change the contained type, we have to early return here rather than
+    // continuing to use monadic operations
+    if (maybePanelTitle)
+        return maybePanelTitle;
+
+    return ensureDict(manifest[sidePanelManifestKey])
+        .and_then([](RetainPtr<NSDictionary<NSString *, id>> sidePanel) { return ensureStringValue(sidePanel.get(), sidePanelPathKey); })
+        .value_or(fallbackPath);
+}
+
+static std::optional<NSDictionary *> getDefaultIconsDictFromExtension(WebExtension& extensions)
+{
+    // FIXME: <https://webkit.org/b/276833> implement this
+    return std::nullopt;
+}
+
+template<typename T>
+static std::optional<T *> nilToNullopt(T *maybeNil)
+{
+    if (maybeNil)
+        return maybeNil;
+    return std::nullopt;
+}
```

Comment:
Move to `CocoaHelpers`, maybe as `toOptional`. But also needs to use `RetainPtr` otherwise this will leak objects.

```suggestion
template<typename T>
static std::optional<RetainPtr<T>> toOptional(T *maybeNil)
{
    if (maybeNil)
        return maybeNil;
    return std::nullopt;
}
```